cluster-autoscaler/scale_down.go

```diff
@@ -197,14 +204,28 @@ func (sd *ScaleDown) TryToScaleDown(nodes []*apiv1.Node, pods []*apiv1.Pod) (Sca
 			glog.V(0).Infof("Scale-down: removing empty node %s", node.Name)
 			simulator.RemoveNodeFromTracker(sd.usageTracker, node.Name, sd.unneededNodes)
 			go func(nodeToDelete *apiv1.Node) {
-				confirmation <- deleteNodeFromCloudProvider(nodeToDelete, sd.context.CloudProvider, sd.context.Recorder)
+				confirmation <- deleteNodeFromCloudProvider(nodeToDelete, sd.context.CloudProvider,
+					sd.context.Recorder, sd.context.ClusterStateRegistry)
 			}(node)
 		}
 		var finalError error
+
+		startTime := time.Now()
 		for range emptyNodes {
-			if err := <-confirmation; err != nil {
-				glog.Errorf("Problem with empty node deletion: %v", err)
-				finalError = err
+			timeElapsed := time.Now().Sub(startTime)
+			timeLeft := MaxCloudProviderNodeDeletionTime - timeElapsed
+			if timeLeft < 0 {
+				finalError = fmt.Errorf("Failed to delete nodes in time")
+				break
+			}
+			select {
+			case err := <-confirmation:
+				if err != nil {
+					glog.Errorf("Problem with empty node deletion: %v", err)
+					finalError = err
+				}
+			case <-time.After(timeLeft):
```

Comment:
Not at this moment, I will be adding cloud provider failure related tests later.